### DISCUSSION TITLE
Bug with connection loosing

### DISCUSSION THREAD

**Discussion by plotn** [dbeaver/dbeaver#16159] [Q&A]
- Operating system (distribution) and version: Win10
- DBeaver version:  21.3.5.202202200327
- Database name and version: Postgres 14

The old bug with connection loosing exists. Synopsys is as follows:

1. I do common work (connection with database is via internet, via vpn);
2. I hibernate windows machine via power menu - not exiting from DBeaver, not turning off vpn;
3. I restore workstation (at the same state). Vpn says it must be reconnected, so I reconnect;
4. I press "refresh" in the tree in DBeaver (refrsh connection), so the connection refreshes normally;
5. I have OLD tab with SQL editor (from yesterday) and a NEW tab - just opened. Connection is 100% the same in both;
6. Trying to make query in NEW tab - all OK;
7. Trying in OLD - notification "Datasource was invalidated" blinks several times. Then "connection was already closed" is written in result tab;
8. Try again - exacltly same behaviour.

This is very very old bug. Logs ar
[LostConn1.txt](https://github.com/dbeaver/dbeaver/files/8168222/LostConn1.txt)
[LostConn2.txt](https://github.com/dbeaver/dbeaver/files/8168224/LostConn2.txt)
e attached.

**Comment by flambchap** [comment]
Hi @plotn. It's expected behavior. The connection is not the same. DBeaver opens a new connection for every new SQL Editor tab by default. You can turn off this in the preferences.

**Reply by plotn** [reply]
@akilovich , loosing the connection in the particular window - this is ok.
But unavailability of restore it - it is not "expected behaviour" ?

**Reply by flambchap** [reply]
I'm sorry, from the steps above, it seemed to me that the connection reestablishes successfully. Could you please make a screencast?

**Reply by plotn** [reply]
Никита, ну какие же "степс абов"??? Посмотри на 7й шаг пожалуйста. 
Ну серьезно, я полгода уже рапортую об этой проблеме (и не я один), а вы все еще думаете - а есть ли она? Ну, если нет, то закрывайте дискуссию да и ладно. 

Давай я переведу написанное - старое окно, которое "со вчера". В нем если запустить запрос - возникает голубая всплывашка возле трея, которая говорит, что коннекшин и правда эстаблишед нормально, причем мигает раз 5. Потом В ЭТОМ САМОМ ОКНЕ внизу, где хотелось бы увидеть результаты запроса выводится надпись, что "соединение закрыто". И все. Хоть 1 раз перезапускай, хоть 10. 
Копирую запрос в новое окно и все работает, старое нужно закрывать, не поможет больше ничего (рефреш всей БД в дереве тоже).

Извини уж, за эмоциональность.

**Reply by flambchap** [reply]
Окей, я перечитал еще раз очень внимательно твое оригинальное сообщение и сообщение выше, и понял, что совсем не так тебя понял :(

Если я теперь правильно понимаю, сообщение про закрытый коннекшен появляется именно в табе, в которой должен быть резалтсет. Если так, то это и правда дичь и бага.

Совершенно не в оправдание нам, но у нас нередко появляются тикеты, которые основаны на непонимании того факта, что по дефолту под каждым редактором есть отдельный конекшен. Я прочитал твой тикет по диагонали и подумал, что это такой же случай. Прошу прощения.

Я правильно понимаю, что в тикете 12559 репорт об этом же? Если так, то я предлагаю перевести обсуждение туда.

**Reply by plotn** [reply]
Да, именно в том самом табе ("вчерашнем"). Т.е. "прокисает" таб и ничего не помогает "вернуть ему коннекшин" (хотя бобер очень старается и что то там восстанавливает, но видимо баг именно в процессе восстановления коннекшина - он хоть и проводится, но каким то образом до исходного таба не дотягивается), хотя "во всём другом" (дерево объектов, другие "свежие" табы) - все работает. 

Тикет конечно же об этом же, все верно.

Спасибо!

**Reply by flambchap** [reply]
@plotn У меня так и не получилось воспроизвести :( А ты кроме VPN пользуешься еще проксей какой-то? А PGBouncer?

**Comment by plotn** [comment]
@akilovich , привет!
Нет, прямое подключение к БД, через VPN.
А логи не помогают понять? Там же видно, как я несколько раз нажимаю "выполнить запрос" и несколько раз получаю:

org.postgresql.util.PSQLException: Соединение уже было закрыто

Может быть отловить это исключение и в этот момент попробовать его переподключить? Т.е. такое ощущение, что то восстановление коннекта, которое мелькает на экране - оно в целом проходит успешно, но обновляет коннект не для данного окна, а "где то еще".

**Reply by flambchap** [reply]
Не хочется чинить вслепую :( Вообще у постгреса весьма специфичный способ хэндлить коннекшены с клиентами, и в бобре чисто ради этого навернут отдельный слой абстракции сверху, и периодически да, баги в связи с этой прекрасной особенностью ПГ. Я попробую привлечь внимание нашего CTO, чтобы он занялся этой проблемой.

**Comment by plotn** [comment]
Как вариант - сделайте мне приватный билд с подобным костылем. Я погоняю, посмотрю как он работает. Если пользы будет больше чем вреда - от этой точки уже можно будет отталкиваться, чтобы дальше думать

Или докидайте логов в этом месте, чтобы проверить нужные гипотезы

**Reply by flambchap** [reply]
Я привлек внимание CTO, заассайнил тикет на него и положил в следующий спринт. Думаю, Серж сможет пофиксить сразу по-человечески.

**Reply by flambchap** [reply]
Тикет 12559. Не кидаю ссылку на тикет, ибо тогда эта дискуссия засветится в самом тикете, а она тут на русском языке :|

**Comment by plotn** [comment]
[DBeaverTableRefresh_log.txt](https://github.com/dbeaver/dbeaver/files/8797053/DBeaverTableRefresh_log.txt)

Никита, вот еще подобная ситуация. В целом все также:
1. Есть объект "соединение", под ним "живет" одна база;
2. Как мы помним, у меня ДБивер открыт "со вчера", при этом я завершал работу виндовс через "гибернацию", после чего переподключал ВПН (когда разгибернировал). ВПН у меня CheckPoint SSL Network Extender$
3. Есть "старое окно" с таблицей (именно с таблицей, не с редактором!!!);
4. Нажимаю правую на соединении, говорю "рефреш". Нажимаю на базе, говорю "рефреш". Все нормально, подключается.
5. Иду на вкладку этой самой таблицы, перехожу в "данные". Нажимаю рефреш (Ф5). Вижу в правом нижнем углу сообщение "Datasource was invalidated", при этом вместо данных вижу "SQL Error [08003]: Соединение уже было закрыто".
6. Логи прикладываю.

**Reply by flambchap** [reply]
Спасибо. Это одна и та же бага, летит опять из класса `org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead`.

**Reply by flambchap** [reply]
Привет @plotn. Пришли, пожалуйста, все логи целиком (https://github.com/dbeaver/dbeaver/wiki/Log-files#log-files) на nikita at jkiss dot org (и там и продолжим общение)

**Comment by MrideaFur** [comment]
Добрейшего денёчка) подскажите, пофиксилась данная проблема? в бобре 23.2.2 наблюдаю её =|

**Reply by plotn** [reply]
мне кажется, что нет. 
Я уже плюнул на все это

**Comment by RiskofStorm** [comment]
Quick fix: disconnect - connect to db

Быстро фиксится отсоединением-присоединением к БД
Забавно что встроенные методы работают, SQL - код нет

ver: 24.2.0.202409011551

**Reply by zabavin** [reply]
версия 25.1.5 фикса так и не вышло :(
